keps/sig-node/20190626-node-shutdown-state.md

```diff
@@ -0,0 +1,253 @@
+* as a cluster administrator I want my stateful workload to failover in case of a node is shutdown without any intervention
+
+* as a developer I would like to rely on a self-healing platform for my stateful workloads
+
+This KEP do not intend to introduce any new API, instead we plan on re-using the Lease API and a mechanism which is similar to leader election (if not the same).
+As stated this KEP is all about finding the right inter-lock between all the actors. The main are:
+
+* Kube-Controller-Manager
+* Kubelets
+* User's workloads
+
+Also Users need to be able to tweak such behavior. To do so, we will introduce a new taint:
+
+```
+taints:
+- key: "cloudprovider.kubernetes.io/node-shutdown"
+  operator: "Exists"
+  effect: "NoSchedule"
+```
+
+we will also deprecate the previous node shutdown taint `node.cloudprovider.kubernetes.io/shutdown` and remove it according to our deprecation
+policy. This enables us to not break out-of-tree clients that were depending on this.
+
+Some users do not want to evict their pods in case of a quick shutdown, they can apply the following toleration:
+
+```
+tolerations:
+- key: "cloudprovider.kubernetes.io/node-failover"
+  operator: "Exists"
+  effect: "NoSchedule"
+  tolerationSeconds: 180
+```
+
+this indicates that a user's workload will failover after 3min.
+
+
+
+#### kube-controller-manager
+
+
+kube-controller-manager needs to provide a controller that:
+
+- Calls the cloud provider
+- Apply the node-shutdown taint
+- removes forcefully the pods
+- Removes the taint
+
+given the current design of our controllers and their following of the single-responsibility pattern, we’ll introduce a new controller called gc_controller under pkg/controller/cloud/podgc.
+
+This controller assumes running on the out-of-tree CCM, to avoid piling features on the deprecated in-tree providers.
+
+This cloud gc_controller watches node objects and when notified (each step describe the possible race condition that could happen)
+
+1. checks the node state via IsShutdownByProviderID(): if there’s a race condition with the cloud provider (i.e. the instance came back see 2.a)
+2. apply a node-shutdown taint
+3. tries to hold node heartbeat lease (Locking is client-side relying on the holder and resourceVersion fields)
+  3.a. fails: backoff and retry if max retries hit -> aborts
+  3.b. success: hold the lease for this node: if the node comes back at this moment it won’t be able to acquire the lease -> pods won’t start
+4. register node as being processed in order to renew the Lease periodically
+5. delete the pod with a 0-grace-period
+7. once 5. returns untaint the node and stop renewing the Lease
+
+
+To avoid being too reactive in 2. we can backoff and not try to acquire the lock immediately to avoid evicting too fast
+
+#### Kubelet
+
+we need also to apply changes to the kubelet, at startup the kubelet will:
+
+1. acquire the Lease:
+  a. fails: do not start any pods, unless the pod tolerates the node shutdown taint
```

Comment:
The node-shutdown taint seems to fulfill two separate purposes:

1. A "normal" NoSchedule taint applied by the controller manager
2. An "implicit" taint the kubelet self-applies to itself when it can't acquire its lease.

I say these two are distinct because, from the node's perspective, one isn't required for the other.  If I read the design correctly, one could choose to just acquire the node's lease and not apply the taint, and it would have the same effect as doing both.  Once concern I have is that if other uses in the future want to use acquiring the node's lease to indicate that the kubelet shouldn't act, we will add lots of "implied" taints for the same behavior, or tolerating "node-shutdown" will come to mean tolerating anytime the kubelet couldn't acquire its lease.

IMO we should separate out these two pieces of functionality.  The "implicit" kubelet taint should be specific to the behavior that caused it, such as `something.kubernetes.io/failed-aquire-lease`.